libraries/Microsoft.Bot.Builder.Dialogs/DialogSet.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Bot.Builder.Dialogs
+{
+    /// <summary>
+    /// A related set of dialogs that can all call each other.
+    /// </summary>
+    /// <typeparam name="C">TurnContext</typeparam>
+    public class DialogSet
+    {
+        private IDictionary<string, Dialog> _dialogs;
+
+        public DialogSet()
+        {
+            _dialogs = new Dictionary<string, Dialog>();
+        }
+
+        /// <summary>
+        /// Adds a new dialog to the set and returns the added dialog.
+        /// </summary>
+        public Dialog Add(string dialogId, Dialog dialog)
+        {
+            if (_dialogs.ContainsKey(dialogId))
+            {
+                throw new Exception($"DialogSet.add(): A dialog with an id of '{dialogId}' already added.");
+            }
+            return _dialogs[dialogId] = dialog;
+        }
+
+        /// <summary>
+        /// Adds a new waterfall to the set and returns the added waterfall.
+        /// </summary>
+        public Waterfall Add(string dialogId, WaterfallStep[] steps)
+        {
+            var waterfall = new Waterfall(steps);
+            Add(dialogId, waterfall);
+            return waterfall;
+        }
+
+        public DialogContext CreateContext(ITurnContext context, object state)
+        {
+            var d = (IDictionary<string, object>)state;
```

Comment:
This is... not ok.

Reply:
See my other comment as its related... We should figure out how in C# you pass in the state bag.

---
In reply to: [181196972](https://github.com/Microsoft/botbuilder-dotnet/pull/408#discussion_r181196972) [](ancestors = 181196972)